from __future__ import annotations
from typing import List, Optional, TYPE_CHECKING, Union

# noinspection PyPackageRequirements
from pyspark.sql.types import StructType, DataType
from spark_auto_mapper_fhir.fhir_types.list import FhirList
from spark_auto_mapper_fhir.fhir_types.integer import FhirInteger
from spark_auto_mapper_fhir.fhir_types.string import FhirString
from spark_auto_mapper_fhir.complex_types.meta import Meta
from spark_auto_mapper_fhir.extensions.extension_base import ExtensionBase
from spark_auto_mapper_fhir.fhir_types.id import FhirId
from spark_auto_mapper_fhir.fhir_types.uri import FhirUri

from spark_auto_mapper_fhir.base_types.fhir_resource_base import FhirResourceBase
from spark_fhir_schemas.r4.resources.molecularsequence import MolecularSequenceSchema

if TYPE_CHECKING:
    pass
    # id_ (id)
    # meta (Meta)
    # implicitRules (uri)
    # language (CommonLanguages)
    from spark_auto_mapper_fhir.value_sets.common_languages import CommonLanguagesCode

    # text (Narrative)
    from spark_auto_mapper_fhir.complex_types.narrative import Narrative

    # contained (ResourceContainer)
    from spark_auto_mapper_fhir.complex_types.resource_container import (
        ResourceContainer,
    )

    # extension (Extension)
    # modifierExtension (Extension)
    # identifier (Identifier)
    from spark_auto_mapper_fhir.complex_types.identifier import Identifier

    # type_ (sequenceType)
    from spark_auto_mapper_fhir.value_sets.sequence_type import SequenceTypeCode

    # coordinateSystem (integer)
    # patient (Reference)
    from spark_auto_mapper_fhir.complex_types.reference import Reference

    # Imports for References for patient
    from spark_auto_mapper_fhir.resources.patient import Patient

    # specimen (Reference)
    # Imports for References for specimen
    from spark_auto_mapper_fhir.resources.specimen import Specimen

    # device (Reference)
    # Imports for References for device
    from spark_auto_mapper_fhir.resources.device import Device

    # performer (Reference)
    # Imports for References for performer
    from spark_auto_mapper_fhir.resources.organization import Organization

    # quantity (Quantity)
    from spark_auto_mapper_fhir.complex_types.quantity import Quantity

    # referenceSeq (MolecularSequence.ReferenceSeq)
    from spark_auto_mapper_fhir.backbone_elements.molecular_sequence_reference_seq import (
        MolecularSequenceReferenceSeq,
    )

    # variant (MolecularSequence.Variant)
    from spark_auto_mapper_fhir.backbone_elements.molecular_sequence_variant import (
        MolecularSequenceVariant,
    )

    # observedSeq (string)
    # quality (MolecularSequence.Quality)
    from spark_auto_mapper_fhir.backbone_elements.molecular_sequence_quality import (
        MolecularSequenceQuality,
    )

    # readCoverage (integer)
    # repository (MolecularSequence.Repository)
    from spark_auto_mapper_fhir.backbone_elements.molecular_sequence_repository import (
        MolecularSequenceRepository,
    )

    # pointer (Reference)
    # Imports for References for pointer
    # structureVariant (MolecularSequence.StructureVariant)
    from spark_auto_mapper_fhir.backbone_elements.molecular_sequence_structure_variant import (
        MolecularSequenceStructureVariant,
    )


# This file is auto-generated by generate_classes so do not edit manually
# noinspection PyPep8Naming
class MolecularSequence(FhirResourceBase):
    """
    MolecularSequence
    molecularsequence.xsd
        Raw data describing a biological sequence.
        If the element is present, it must have either a @value, an @id, or extensions
    """

    # noinspection PyPep8Naming
    def __init__(
        self,
        *,
        use_date_for: Optional[List[str]] = None,
        id_: Optional[FhirId] = None,
        meta: Optional[Meta] = None,
        implicitRules: Optional[FhirUri] = None,
        language: Optional[CommonLanguagesCode] = None,
        text: Optional[Narrative] = None,
        contained: Optional[FhirList[ResourceContainer]] = None,
        extension: Optional[FhirList[ExtensionBase]] = None,
        modifierExtension: Optional[FhirList[ExtensionBase]] = None,
        identifier: Optional[FhirList[Identifier]] = None,
        type_: Optional[SequenceTypeCode] = None,
        coordinateSystem: FhirInteger,
        patient: Optional[Reference[Patient]] = None,
        specimen: Optional[Reference[Specimen]] = None,
        device: Optional[Reference[Device]] = None,
        performer: Optional[Reference[Organization]] = None,
        quantity: Optional[Quantity] = None,
        referenceSeq: Optional[MolecularSequenceReferenceSeq] = None,
        variant: Optional[FhirList[MolecularSequenceVariant]] = None,
        observedSeq: Optional[FhirString] = None,
        quality: Optional[FhirList[MolecularSequenceQuality]] = None,
        readCoverage: Optional[FhirInteger] = None,
        repository: Optional[FhirList[MolecularSequenceRepository]] = None,
        pointer: Optional[FhirList[Reference[MolecularSequence]]] = None,
        structureVariant: Optional[FhirList[MolecularSequenceStructureVariant]] = None,
    ) -> None:
        """
            Raw data describing a biological sequence.
            If the element is present, it must have either a @value, an @id, or extensions

            :param id_: The logical id of the resource, as used in the URL for the resource. Once
        assigned, this value never changes.
            :param meta: The metadata about the resource. This is content that is maintained by the
        infrastructure. Changes to the content might not always be associated with
        version changes to the resource.
            :param implicitRules: A reference to a set of rules that were followed when the resource was
        constructed, and which must be understood when processing the content. Often,
        this is a reference to an implementation guide that defines the special rules
        along with other profiles etc.
            :param language: The base language in which the resource is written.
            :param text: A human-readable narrative that contains a summary of the resource and can be
        used to represent the content of the resource to a human. The narrative need
        not encode all the structured data, but is required to contain sufficient
        detail to make it "clinically safe" for a human to just read the narrative.
        Resource definitions may define what content should be represented in the
        narrative to ensure clinical safety.
            :param contained: These resources do not have an independent existence apart from the resource
        that contains them - they cannot be identified independently, and nor can they
        have their own independent transaction scope.
            :param extension: May be used to represent additional information that is not part of the basic
        definition of the resource. To make the use of extensions safe and manageable,
        there is a strict set of governance  applied to the definition and use of
        extensions. Though any implementer can define an extension, there is a set of
        requirements that SHALL be met as part of the definition of the extension.
            :param modifierExtension: May be used to represent additional information that is not part of the basic
        definition of the resource and that modifies the understanding of the element
        that contains it and/or the understanding of the containing element's
        descendants. Usually modifier elements provide negation or qualification. To
        make the use of extensions safe and manageable, there is a strict set of
        governance applied to the definition and use of extensions. Though any
        implementer is allowed to define an extension, there is a set of requirements
        that SHALL be met as part of the definition of the extension. Applications
        processing a resource are required to check for modifier extensions.

        Modifier extensions SHALL NOT change the meaning of any elements on Resource
        or DomainResource (including cannot change the meaning of modifierExtension
        itself).
            :param identifier: A unique identifier for this particular sequence instance. This is a FHIR-
        defined id.
            :param type_: Amino Acid Sequence/ DNA Sequence / RNA Sequence.
            :param coordinateSystem: Whether the sequence is numbered starting at 0 (0-based numbering or
        coordinates, inclusive start, exclusive end) or starting at 1 (1-based
        numbering, inclusive start and inclusive end).
            :param patient: The patient whose sequencing results are described by this resource.
            :param specimen: Specimen used for sequencing.
            :param device: The method for sequencing, for example, chip information.
            :param performer: The organization or lab that should be responsible for this result.
            :param quantity: The number of copies of the sequence of interest. (RNASeq).
            :param referenceSeq: A sequence that is used as a reference to describe variants that are present
        in a sequence analyzed.
            :param variant: The definition of variant here originates from Sequence ontology ([variant_of]
        (http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This
        element can represent amino acid or nucleic sequence change(including
        insertion,deletion,SNP,etc.)  It can represent some complex mutation or
        segment variation with the assist of CIGAR string.
            :param observedSeq: Sequence that was observed. It is the result marked by referenceSeq along with
        variant records on referenceSeq. This shall start from
        referenceSeq.windowStart and end by referenceSeq.windowEnd.
            :param quality: An experimental feature attribute that defines the quality of the feature in a
        quantitative way, such as a phred quality score ([SO:0001686](http://www.seque
        nceontology.org/browser/current_svn/term/SO:0001686)).
            :param readCoverage: Coverage (read depth or depth) is the average number of reads representing a
        given nucleotide in the reconstructed sequence.
            :param repository: Configurations of the external repository. The repository shall store target's
        observedSeq or records related with target's observedSeq.
            :param pointer: Pointer to next atomic sequence which at most contains one variant.
            :param structureVariant: Information about chromosome structure variation.
        """
        super().__init__(
            resourceType="MolecularSequence",
            id_=id_,
            meta=meta,
            implicitRules=implicitRules,
            language=language,
            text=text,
            contained=contained,
            extension=extension,
            modifierExtension=modifierExtension,
            identifier=identifier,
            type_=type_,
            coordinateSystem=coordinateSystem,
            patient=patient,
            specimen=specimen,
            device=device,
            performer=performer,
            quantity=quantity,
            referenceSeq=referenceSeq,
            variant=variant,
            observedSeq=observedSeq,
            quality=quality,
            readCoverage=readCoverage,
            repository=repository,
            pointer=pointer,
            structureVariant=structureVariant,
        )

        self.use_date_for = use_date_for

    def get_schema(
        self, include_extension: bool, extension_fields: Optional[List[str]] = None
    ) -> Optional[Union[StructType, DataType]]:
        return MolecularSequenceSchema.get_schema(
            include_extension=include_extension,
            extension_fields=extension_fields,
            use_date_for=self.use_date_for,
        )
